### PR TITLE
Support filter capability for Tenant Management API

### PR DESCRIPTION
**Pull request by hwupathum**
### Purpose 

This pull request introduces support for filtering tenant records in the tenant management API. The key changes include defining new filter attributes and operations, updating error messages, and implementing the filtering logic.

### Related issue

- https://github.com/wso2/product-is/issues/21374

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java [hunk 1]
```diff
@@ -52,6 +63,12 @@ public enum ErrorMessage {
         ERROR_CODE_PARTIALLY_CREATED_OR_UPDATED("TM-60021",
                 "Tenant creation / update was completed with errors.",
                 "Tenant creation / update was completed with error: %s"),
+        ERROR_CODE_INVALID_FILTER_FORMAT("TM-60022", "Invalid format used for filtering.",
+                "Filter needs to be in the format <attribute>+<operation>+<value>."),
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.common/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/common/TenantManagementConstants.java:67]: How about following ?
Invalid filter format. , The filter must follow the format: <attribute>+<operation>+<value>.

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/tenant/management/v1/TenantsApi.java [hunk 2]
```diff
@@ -238,13 +238,13 @@ public Response isDomainExist(@ApiParam(value = "tenant domain",required=true) @
...
+    public Response retrieveTenants(    @Valid @Min(0)@ApiParam(value = "Maximum number of records to return.")  @QueryParam("limit") Integer limit,     @Valid @Min(0)@ApiParam(value = "Number of records to skip for pagination.")  @QueryParam("offset") Integer offset,     @Valid@ApiParam(value = "Define the order in which the retrieved tenants should be sorted.", allowableValues="asc, desc")  @QueryParam("sortOrder") String sortOrder,     @Valid@ApiParam(value = "Attribute by which the retrieved records should be sorted. Currently sorting through <b>domainName</b> only supported.")  @QueryParam("sortBy") String sortBy,     @Valid@ApiParam(value = "Condition to filter the retrieval of records. Supports 'sw', 'co', 'ew' and 'eq' operations. Currently filtering through <b>domainName</b> only supported. E.g. /tenants?filter=domainName+sw+wso2. ")  @QueryParam("filter") String filter) {
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/tenant/management/v1/TenantsApi.java:247]: How about 

Currently filtering through <b>domainName</b> only supported. E.g. /tenants?filter=domainName+sw+wso2.

---> 
Currently, filtering is only supported via the <b>domainName</b>. For example: /tenants?filter=domainName+sw+wso2.

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java [hunk 3]
```diff
@@ -45,6 +45,7 @@
 import org.wso2.carbon.stratos.common.exception.TenantManagementServerException;
 import org.wso2.carbon.stratos.common.exception.TenantMgtException;
 import org.wso2.carbon.stratos.common.util.ClaimsMgtUtil;
+import org.wso2.carbon.tenant.mgt.services.TenantMgtImpl;
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java:48]: Please double check this import. Smells like cyclic dependency on imp usage.

We can reconsider the constant location TenantMgtImpl.DOMAIN_NAME or somewhere else depending on the above decision.

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/resources/tenant-management.yaml [hunk 4]
```diff
@@ -497,10 +493,9 @@ components:
       name: filter
       required: false
       description: >
-        Condition to filter the retrival of records. Supports 'sw', 'co', 'ew'
-        and 'eq' operations and also complex queries with 'and' operations. E.g.
-        /tenants?filter=domain+sw+"wso2". _<b>This
-        option is not yet supported.<b>_
+        Condition to filter the retrieval of records. Supports 'sw', 'co', 'ew' and 'eq' operations. Currently filtering
+        through <b>domainName</b> only supported.
```
Comment (indeewari) [inline components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/resources/tenant-management.yaml:497]: How about
Condition to filter the retrieval of records. Supports 'sw', 'co', 'ew' and 'eq' operations. Currently, filtering is supported only by the <b>domainName</b>.

File: components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java (hunk 3 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.api.server.tenant.management/org.wso2.carbon.identity.api.server.tenant.management.v1/src/main/java/org/wso2/carbon/identity/api/server/tenant/management/v1/core/ServerTenantManagementService.java]: Moved to https://github.com/wso2/carbon-multitenancy/pull/279

**Review by indeewari — APPROVED** [review]

**Pull request merged**